pallets/gear/src/migrations.rs

```diff
@@ -0,0 +1,372 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Config, Pallet, QueueOf};
+use common::{
+    event::{MessageWokenSystemReason, SystemReason},
+    storage::Queue,
+    Origin,
+};
+use frame_support::{
+    traits::{Get, GetStorageVersion, OnRuntimeUpgrade},
+    weights::Weight,
+};
+use gear_core::ids::MessageId;
+use sp_std::{marker::PhantomData, vec::Vec};
+
+#[cfg(feature = "try-runtime")]
+use {
+    common::storage::IterableMap,
+    parity_scale_codec::{Decode, Encode},
+    sp_runtime::TryRuntimeError,
+};
+
+pub struct MigrateWaitingInitList<T>(PhantomData<T>);
+
+impl<T> OnRuntimeUpgrade for MigrateWaitingInitList<T>
+where
+    T: Config + pallet_gear_program::Config,
+    T::AccountId: Origin,
+{
+    fn on_runtime_upgrade() -> Weight {
+        let current = pallet_gear_program::Pallet::<T>::current_storage_version();
+        let onchain = pallet_gear_program::Pallet::<T>::on_chain_storage_version();
+
+        log::info!(
+            "🚚 Running migration with current storage version {current:?} / onchain {onchain:?}"
+        );
+
+        // 1 read for the on-chain storage version
+        let mut weight = T::DbWeight::get().reads(1);
+
+        if current == 4 && onchain == 3 {
+            waiting_init_list::WaitingInitStorage::<T>::translate(
+                |program_id, messages: Vec<MessageId>| {
+                    // read and remove an element
+                    weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+                    for message_id in messages {
+                        if let Some(dispatch) = Pallet::<T>::wake_dispatch(
+                            program_id,
+                            message_id,
+                            MessageWokenSystemReason::WaitInitListMigration.into_reason(),
+                        ) {
+                            // remove from waitlist
+                            weight = weight.saturating_add(T::DbWeight::get().writes(1));
+
+                            QueueOf::<T>::queue(dispatch).unwrap_or_else(|e| {
+                                log::error!("Message queue corrupted! {:?}", e)
+                            });
+
+                            // push to queue
+                            weight = weight.saturating_add(T::DbWeight::get().writes(1));
+                        }
+                    }
+
+                    None
+                },
+            );
+
+            current.put::<pallet_gear_program::Pallet<T>>();
+
+            log::info!("Successfully migrated storage");
+        } else {
+            log::info!("❌ Migration did not execute. This probably should be removed");
+        }
+
+        weight
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
```

Comment:
please make pre_upgrade and post_upgrade dependent on storage version
ref: https://github.com/gear-tech/gear/pull/3869/commits/6253cea6c1a57337f25dd1fa295e250444b0525d

Reply:
I don't get why

Reply:
because latest versions of try-runtime cli we use to check runtime migrations has step of checking indempotency of the migrations and fail without this check